AsyncDisplayKit/ASMapNode.mm

```diff
@@ -0,0 +1,210 @@
+      _options.size = self.calculatedSize;
+      _snapshotter = [[MKMapSnapshotter alloc] initWithOptions:_options];
+  }
+}
+
+- (BOOL)isLiveMap
+{
+  ASDN::MutexLocker l(_propertyLock);
+  return _liveMap;
+}
+
+- (void)setLiveMap:(BOOL)liveMap
+{
+  ASDN::MutexLocker l(_propertyLock);
+  if (liveMap == _liveMap) {
+    return;
+  }
+  _liveMap = liveMap;
+  liveMap ? [self addLiveMap] : [self removeLiveMap];
+}
+
+
+- (BOOL)needsMapReloadOnBoundsChange
+{
+  ASDN::MutexLocker l(_propertyLock);
+  return _needsMapReloadOnBoundsChange;
+}
+
+- (void)setNeedsMapReloadOnBoundsChange:(BOOL)needsMapReloadOnBoundsChange
+{
+  ASDN::MutexLocker l(_propertyLock);
+  _needsMapReloadOnBoundsChange = needsMapReloadOnBoundsChange;
+}
+
+- (void)fetchData
+{
+  [super fetchData];
+  if (_liveMap && !_mapNode) {
+    [self addLiveMap];
+  }
+  else {
+    [self setUpSnapshotter];
+    [self takeSnapshot];
+  }
+}
+
+- (void)clearFetchedData
+{
+  [super clearFetchedData];
+  [self removeLiveMap];
+}
+
+- (void)takeSnapshot
+{
+  if (!_snapshotter.isLoading) {
+    [_snapshotter startWithCompletionHandler:^(MKMapSnapshot *snapshot, NSError *error) {
+      if (!error) {
+        UIImage *image = snapshot.image;
+        CGRect finalImageRect = CGRectMake(0, 0, image.size.width, image.size.height);
+        
+        UIGraphicsBeginImageContextWithOptions(image.size, YES, image.scale);
+        [image drawAtPoint:CGPointMake(0, 0)];
+        
+        if (_annotations.count > 0 ) {
+          // Get a standard annotation view pin. Future implementations should use a custom annotation image property.
+          MKAnnotationView *pin = [[MKPinAnnotationView alloc] initWithAnnotation:nil reuseIdentifier:@""];
+          UIImage *pinImage = pin.image;
+          for (id<MKAnnotation>annotation in _annotations)
+          {
+            CGPoint point = [snapshot pointForCoordinate:annotation.coordinate];
+            if (CGRectContainsPoint(finalImageRect, point))
+            {
+              CGPoint pinCenterOffset = pin.centerOffset;
+              point.x -= pin.bounds.size.width / 2.0;
+              point.y -= pin.bounds.size.height / 2.0;
+              point.x += pinCenterOffset.x;
+              point.y += pinCenterOffset.y;
+              [pinImage drawAtPoint:point];
+            }
+          }
+        }
+        
+        UIImage *finalImage = UIGraphicsGetImageFromCurrentImageContext();
+        UIGraphicsEndImageContext();
+        self.image = finalImage;
+      }
+    }];
+  }
+}
+
+- (void)resetSnapshotter
+{
+  if (!_snapshotter.isLoading) {
+    _options.size = self.calculatedSize;
+    _snapshotter = [[MKMapSnapshotter alloc] initWithOptions:_options];
+  }
+}
+
+#pragma mark - Action
+- (void)addLiveMap
+{
+  if (self.isNodeLoaded && !_mapNode) {
```

Comment:
I think we should improve this method in a followup diff.

- Does it work correctly to set isLiveMap before the node is loaded?  This code path bails, and I couldn't readily find what else creates the _mapView on node loading if not this (e.g. there is no -didLoad method).
- Add spaces after the "alloc]" calls below.
- Let's test the memory behavior here.  It is possible there is a strong reference cycle.  You may need to create a "weakSelf" local variable and access things like mapDelegate on it; there is an implicit retain / reference to self in order to access the instance variables directly within the block.  Even after the view load block is called, it will be retained by the node.  We should test a sample app with an ASMapNode that goes to the fully liveMap state, and then deallocate the ASMapNode, and check with Instruments Leaks + Allocations that it is destroyed.
- It might be best to just init with CGRectZero and call setNeedsLayout on weakSelf at the end.  It might be equally good to access weakSelf.bounds (make a local variable CGSize boundsSize = weakSelf.bounds.size in this case, to avoid calling bounds property twice) - and replace calculatedSize with that - because nothing requires the node to be positioned / sized equal to the calculatedSize.